distributed/shuffle/_rechunk.py

```diff
@@ -252,3 +264,226 @@ def split_axes(old: ChunkedAxes, new: ChunkedAxes) -> SplitAxes:
+            partitions_of[addr].append(part)
+        self.partitions_of = dict(partitions_of)
+        self.worker_for = worker_for
+        self.split_axes = split_axes(old, new)
+
+    async def _receive(self, data: list[tuple[NDIndex, bytes]]) -> None:
+        self.raise_if_closed()
+
+        filtered = []
+        for d in data:
+            id, payload = d
+            if id in self.received:
+                continue
+            filtered.append(payload)
+            self.received.add(id)
+            self.total_recvd += sizeof(d)
+        del data
+        if not filtered:
+            return
+        try:
+            shards = await self.offload(self._repartition_shards, filtered)
+            del filtered
+            await self._write_to_disk(shards)
+        except Exception as e:
+            self._exception = e
+            raise
+
+    def _repartition_shards(self, data: list[bytes]) -> dict[NDIndex, bytes]:
+        repartitioned: defaultdict[
+            NDIndex, list[tuple[NDIndex, np.ndarray]]
+        ] = defaultdict(list)
+        for buffer in data:
+            for id, shard in pickle.loads(buffer):
+                repartitioned[id].append(shard)
+        return {k: pickle.dumps(v) for k, v in repartitioned.items()}
+
+    async def add_partition(self, data: np.ndarray, partition_id: NDIndex) -> int:
+        self.raise_if_closed()
+        if self.transferred:
+            raise RuntimeError(f"Cannot add more partitions to {self}")
+
+        def _() -> dict[str, tuple[NDIndex, bytes]]:
+            """Return a mapping of worker addresses to a tuple of input partition
+            IDs and shard data.
+
+
+            TODO: Overhaul!
+            As shard data, we serialize the payload together with the sub-index of the
+            slice within the new chunk. To assemble the new chunk from its shards, it
+            needs the sub-index to know where each shard belongs within the chunk.
+            Adding the sub-index into the serialized payload on the sender allows us to
+            write the serialized payload directly to disk on the receiver.
+            """
+            out: dict[
+                str, list[tuple[NDIndex, tuple[NDIndex, np.ndarray]]]
+            ] = defaultdict(list)
+            from itertools import product
+
+            ndsplits = product(
+                *(axis[i] for axis, i in zip(self.split_axes, partition_id))
+            )
+
+            for ndsplit in ndsplits:
+                chunk_index, shard_index, ndslice = zip(*ndsplit)
+                out[self.worker_for[chunk_index]].append(
+                    (chunk_index, (shard_index, data[ndslice]))
+                )
+            return {k: (partition_id, pickle.dumps(v)) for k, v in out.items()}
+
+        out = await self.offload(_)
+        await self._write_to_comm(out)
+        return self.run_id
+
+    async def get_output_partition(
+        self, partition_id: NDIndex, key: str, meta: pd.DataFrame | None = None
```

Comment:
As above, if this function were generic over meta as `_T_partition_type` then `meta` could be `meta : np.ndarray | None` here, and rechunking wouldn't require pandas for this type annotation.